src/liteclient.cc

```diff
@@ -105,10 +94,28 @@ LiteClient::LiteClient(Config& config_in)
     throw std::runtime_error("Unsupported package manager type: " + config.pacman.type);
   }
 
-  writeCurrentTarget(current_target);
-  if (target_installation_result.result_code != data::ResultCode::Numeric::kAlreadyProcessed) {
-    notifyInstallFinished(target_been_applied, target_installation_result);
+  {
+    // finalize a pending update if any
+    boost::optional<Uptane::Target> pending_target;
+    storage->loadInstalledVersions("", nullptr, &pending_target);
+
+    if (!!pending_target) {
+      data::InstallationResult update_finalization_result = package_manager_->finalizeInstall(*pending_target);
+      if (update_finalization_result.isSuccess()) {
+        LOG_INFO << "Marking target install complete for: " << *pending_target;
+        storage->saveInstalledVersion("", *pending_target, InstalledVersionUpdateMode::kCurrent);
+      }
```

Comment:
it looks like there used to be an "else" hanging of `!!pending_target`
~~~
} else {
        if (bootloader.rebootDetected()) {
          std::string err = "Expected to boot on " + target.sha256Hash() + " buf found " + current_hash +
                            ", system might have experienced a rollback";
          LOG_ERROR << err;
          storage.saveInstalledVersion("", target, InstalledVersionUpdateMode::kNone);
          bootloader.rebootFlagClear();
          ir.result_code = data::ResultCode::Numeric::kInstallFailed;
          ir.description = err;
        } else {
          // Update still pending as no reboot was detected
          ir.result_code = data::ResultCode::Numeric::kNeedCompletion;
        }   
        // Installation was not successful
        return std::tie(*current_version, target, ir);
      }   
~~~

I think that's kind of `package_manager_->finalizeInstall(*pending_target);` but slightly different? Not sure if its functionally different.

Reply:
conceptually the same, but OstreeManager::finalizeInstall does it correctly and is well tested, the aklite actually used it originally when it used sotauptaneclient.
My version LiteClient::finalizeIfNeeded() has a tiny issue, it actually doesn't verify if a reboot has happened if there is a pending version, just simply checks the current hash which can be applied manually by doing `ostree admin deploy` without reboot. Also, an installation finalization should be done by an entity that actually performs installation, i.e. package manager in our case.

Reply:
ack. i'd like to have @ricardosalveti take a look at this snippet since he understands the rollback stuff more than i do.

Reply:
There is one difference that I could notice is that on the new path we don't call storage.saveInstalledVersion("", target, InstalledVersionUpdateMode::kNone); anymore, and not sure if that can potentially cause an issue on the rollback logic.

Reply:
It's kind of redundant because this call doesn't change the existing record for the given target. What important for the rollback functionality is the flag `was_installed`. It should be set to 1/true only for those Targets that were successfully installed and we do it either in LiteClient::install() for updates that do not require a reboot and in LiteClient::LiteClient() if a reboot is required, reboot happened and the finalization was successful. So, if Target finalization was not successful we just simply don't do anything with the Target record in DB because its `was_installed` is set to 0 anyway.